Handle pandas deprecation warnings for `to_pydatetime` and `apply`

phofl opened the pull request:
- [x] xref #10089
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

phofl commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -3575,10 +3581,12 @@ def return_df2(x):
     with warnings.catch_warnings():
         warnings.simplefilter("ignore", UserWarning)
-        result = ddf.x.apply(return_df2)
+        with pytest.warns(FutureWarning, match="Returning a DataFrame"):
```
Want to ignore the warnings here for now. A couple of cases aren't raising warnings yet (which seems buggy).

jrbourbeau commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -3546,6 +3552,7 @@ def test_autocorr():
 
 
 def test_apply_infer_columns():
+    warning = FutureWarning if PANDAS_GT_210 else None
```
Nit: I would move this definition further down to just before it's used. This will, I think, make it easier for future readers to reason about.

phofl replied:
Updated

jrbourbeau commented:
Also, I forgot to mentioned that our `upstream` build, which installs the nightly `pandas` wheel, don't run on PRs by default. You can opt in to running the `upstream` build on PRs if you include the phrase `test-upstream` in a commit message or add the `upstream` label (I've gone ahead and done that here)

phofl commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -3150,13 +3149,20 @@ def test_apply():
+        # inference
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", UserWarning)
+            assert_eq(ddf.x.pipe(func), df.x.pipe(func))
```
Ok, wasn't sure how to handle this, reverted the changed

phofl commented:
Thx, did not consider our internal warnings. I updated the pr to catch all warnings during the execution phase. Also adjusted one test each to check that we don't show any warnings during the compute stage.  Let me know what you think

phofl commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -3150,13 +3149,20 @@ def test_apply():
-    assert_eq(ddf.x.apply(func, meta=[(0, int), (1, int)]), df.x.apply(func))
+    with pytest.warns(warning, match="Returning a DataFrame"):
+        ddf_result = ddf.x.apply(func, meta=[(0, int), (1, int)])
+        pdf_result = df.x.apply(func)
```
Updated

jrbourbeau reviewed (approved):
Okay, looks like `test_dt_accessor` had a minor issue which I pushed a small commit for (hope that's okay). Will merge this in once CI finishes. 

There are new `upstream` failures like 

```
dask/dataframe/io/tests/test_parquet.py::test_columns_index[pyarrow-fastparquet]: AssertionError: Index are different

Attribute "inferred_type" are different
[left]:  empty
[right]: integer
```

but those are totally unrelated to the changes here

phofl commented:
Thx, not sure what's going on with the match functionality, but could not reproduce this locally.

I saw these earlier and looked into it, needs a fix in fastparquet. pandas merged a small change there yesterday to make Index construction consistent that is causing this here.

Will open a pr to ignore for now tomorrow
